# Unit 12 - Tales from the Crypto

---


## 1. Sentiment Analysis

Use the [newsapi](https://newsapi.org/) to pull the latest news articles for Bitcoin and Ethereum and create a DataFrame of sentiment scores for each coin.

Use descriptive statistics to answer the following questions:
1. Which coin had the highest mean positive score?
2. Which coin had the highest negative score?
3. Which coin had the highest positive score?

In [1]:
# Initial imports
import os
import pandas as pd
from dotenv import load_dotenv
import nltk as nltk
nltk.download('vader_lexicon')
from nltk.sentiment.vader import SentimentIntensityAnalyzer
analyzer = SentimentIntensityAnalyzer()

%matplotlib inline

[nltk_data] Downloading package vader_lexicon to
[nltk_data]     /Users/bennyantony/nltk_data...


In [4]:
# Read your api key environment variable
# Load .env enviroment variables
load_dotenv()

#now get the news API key from the .env file
newsapi_key = os.getenv("newsapi_key")


In [5]:

from newsapi import NewsApiClient
# Create a newsapi client
newsapi = NewsApiClient(api_key=newsapi_key)

In [13]:
# Fetch the Bitcoin news articles
all_bitcoin_articles = newsapi.get_everything(q='bitcoin',
                                             language="en",
                                            page_size=100,
                                            sort_by="relevancy")


print(f"Total articles about Bitcoin: {all_bitcoin_articles['totalResults']}")

Total articles about Bitcoin: 10773


In [14]:
# Fetch the Ethereum news articles
all_ethereum_articles = newsapi.get_everything(q='ethereum',
                                            language="en",
                                            page_size=100,
                                            sort_by="relevancy")

print(f"Total articles about Ethereum: {all_ethereum_articles['totalResults']}")

Total articles about Ethereum: 4300


In [21]:
# Create the Bitcoin sentiment scores DataFrame
bitcoin_senti = []

for article in all_bitcoin_articles["articles"]:
        text = article["content"]
        date = article["publishedAt"][:10]
        sentiment = analyzer.polarity_scores(text)
        compound = sentiment["compound"]
        pos = sentiment["pos"]
        neu = sentiment["neu"]
        neg = sentiment["neg"]
        
        bitcoin_senti.append({
            "text": text,
            "date": date,
            "compound": compound,
            "positive": pos,
            "negative": neg,
            "neutral": neu
            
        })
df_bitcoin = pd.DataFrame(bitcoin_senti)
df_bitcoin.head()


compound        date  negative  neutral  positive  \
0    0.3818  2021-05-12      0.00    0.929     0.071   
1    0.2960  2021-05-07      0.00    0.942     0.058   
2    0.0000  2021-04-27      0.00    1.000     0.000   
3    0.7506  2021-04-20      0.00    0.829     0.171   
4    0.4939  2021-05-12      0.05    0.816     0.134   

                                                text  
0  Just weeks after Tesla started accepting Bitco...  
1  Shares of Square are up more than 6% today aft...  
2  Tesla’s relationship with bitcoin is not a dal...  
3  Cryptocurrency continues to gain mainstream ac...  
4  Image: Tesla\r\n\n \n\n Tesla has stopped acce...

In [22]:
# Create the Ethereum sentiment scores DataFrame
ethereum_senti = []

for article in all_ethereum_articles["articles"]:
        text = article["content"]
        date = article["publishedAt"][:10]
        sentiment = analyzer.polarity_scores(text)
        compound = sentiment["compound"]
        pos = sentiment["pos"]
        neu = sentiment["neu"]
        neg = sentiment["neg"]
        
        ethereum_senti.append({
            "text": text,
            "date": date,
            "compound": compound,
            "positive": pos,
            "negative": neg,
            "neutral": neu
            
        })
df_ethereum = pd.DataFrame(ethereum_senti)
df_ethereum.head()

compound        date  negative  neutral  positive  \
0    0.2263  2021-05-12      0.00    0.940     0.060   
1   -0.4019  2021-05-04      0.15    0.778     0.072   
2    0.7506  2021-04-20      0.00    0.829     0.171   
3    0.0258  2021-04-20      0.00    0.966     0.034   
4    0.4019  2021-05-15      0.00    0.917     0.083   

                                                text  
0  Vitalik Buterin, the creator of Ethereum, on W...  
1  Their investors call them disruptive innovator...  
2  Cryptocurrency continues to gain mainstream ac...  
3  Venmo is jumping aboard the cryptocurrency ban...  
4  Solana isn’t known yet outside of the crypto c...

In [23]:
# Describe the Bitcoin Sentiment
df_bitcoin.describe()

compound    negative     neutral    positive
count  100.000000  100.000000  100.000000  100.000000
mean     0.124340    0.034480    0.899030    0.066520
std      0.392952    0.048393    0.087723    0.073219
min     -0.726900    0.000000    0.662000    0.000000
25%      0.000000    0.000000    0.837500    0.000000
50%      0.000000    0.000000    0.907500    0.057500
75%      0.421500    0.068500    1.000000    0.108000
max      0.845500    0.219000    1.000000    0.275000

In [24]:
# Describe the Ethereum Sentiment
df_ethereum.describe()

compound    negative     neutral   positive
count  100.000000  100.000000  100.000000  100.00000
mean     0.133908    0.025920    0.914370    0.05971
std      0.367237    0.050516    0.078769    0.06712
min     -0.918600    0.000000    0.694000    0.00000
25%      0.000000    0.000000    0.854000    0.00000
50%      0.038700    0.000000    0.927000    0.04700
75%      0.426225    0.047250    1.000000    0.09200
max      0.827100    0.289000    1.000000    0.25600

### Questions:

Q: Which coin had the highest mean positive score?

A: Bitcoin(0.066520) has highest score compared to Ethereum(0.05971)

Q: Which coin had the highest compound score?

A: Bitcoin(0.845500) has highest score compared to Ethereum(0.05971)

Q. Which coin had the highest positive score?

A: 

---

## 2. Natural Language Processing
---
###   Tokenizer

In this section, you will use NLTK and Python to tokenize the text for each coin. Be sure to:
1. Lowercase each word.
2. Remove Punctuation.
3. Remove Stopwords.

In [10]:
from nltk.tokenize import word_tokenize, sent_tokenize
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer, PorterStemmer
from string import punctuation
import re

In [11]:
# Instantiate the lemmatizer
# YOUR CODE HERE!

# Create a list of stopwords
# YOUR CODE HERE!

# Expand the default stopwords list if necessary
# YOUR CODE HERE!

In [12]:
# Complete the tokenizer function
def tokenizer(text):
    """Tokenizes text."""
    
    # Remove the punctuation from text

   
    # Create a tokenized list of the words
    
    
    # Lemmatize words into root words

   
    # Convert the words to lowercase
    
    
    # Remove the stop words
    
    
    return tokens

In [13]:
# Create a new tokens column for Bitcoin
# YOUR CODE HERE!

In [14]:
# Create a new tokens column for Ethereum
# YOUR CODE HERE!

---

### NGrams and Frequency Analysis

In this section you will look at the ngrams and word frequency for each coin. 

1. Use NLTK to produce the n-grams for N = 2. 
2. List the top 10 words for each coin. 

In [15]:
from collections import Counter
from nltk import ngrams

In [16]:
# Generate the Bitcoin N-grams where N=2
# YOUR CODE HERE!

In [17]:
# Generate the Ethereum N-grams where N=2
# YOUR CODE HERE!

In [18]:
# Function token_count generates the top 10 words for a given coin
def token_count(tokens, N=3):
    """Returns the top N tokens from the frequency count"""
    return Counter(tokens).most_common(N)

In [19]:
# Use token_count to get the top 10 words for Bitcoin
# YOUR CODE HERE!

In [20]:
# Use token_count to get the top 10 words for Ethereum
# YOUR CODE HERE!

---

### Word Clouds

In this section, you will generate word clouds for each coin to summarize the news for each coin

In [21]:
from wordcloud import WordCloud
import matplotlib.pyplot as plt
plt.style.use('seaborn-whitegrid')
import matplotlib as mpl
mpl.rcParams['figure.figsize'] = [20.0, 10.0]

In [22]:
# Generate the Bitcoin word cloud
# YOUR CODE HERE!

In [23]:
# Generate the Ethereum word cloud
# YOUR CODE HERE!

---
## 3. Named Entity Recognition

In this section, you will build a named entity recognition model for both Bitcoin and Ethereum, then visualize the tags using SpaCy.

In [24]:
import spacy
from spacy import displacy

In [25]:
# Download the language model for SpaCy
# !python -m spacy download en_core_web_sm

In [26]:
# Load the spaCy model
nlp = spacy.load('en_core_web_sm')

---
### Bitcoin NER

In [27]:
# Concatenate all of the Bitcoin text together
# YOUR CODE HERE!

In [28]:
# Run the NER processor on all of the text
# YOUR CODE HERE!

# Add a title to the document
# YOUR CODE HERE!

In [29]:
# Render the visualization
# YOUR CODE HERE!

In [30]:
# List all Entities
# YOUR CODE HERE!

---

### Ethereum NER

In [31]:
# Concatenate all of the Ethereum text together
# YOUR CODE HERE!

In [32]:
# Run the NER processor on all of the text
# YOUR CODE HERE!

# Add a title to the document
# YOUR CODE HERE!

In [33]:
# Render the visualization
# YOUR CODE HERE!

In [34]:
# List all Entities
# YOUR CODE HERE!

---